src/compiler/checker.ts

```diff
@@ -10611,13 +10608,20 @@ namespace ts {
             return t => t === source ? target : baseMapper(t);
         }
 
-        function wildcardMapper(type: Type) {
+        function permissiveMapper(type: Type) {
             return type.flags & TypeFlags.TypeParameter ? wildcardType : type;
         }
 
+        function getRestrictiveTypeParameter(tp: TypeParameter) {
+            return !tp.constraint ? tp : tp.restrictiveInstantiation || (tp.restrictiveInstantiation = createTypeParameter(tp.symbol));

```

Comment:
This should be
```ts

        function getRestrictiveTypeParameter(tp: TypeParameter) {
            return tp.constraint === noConstraintType ? tp : tp.restrictiveInstantiation || (
                tp.restrictiveInstantiation = createTypeParameter(tp.symbol),
                (tp.restrictiveInstantiation as TypeParameter).constraint = noConstraintType,
                tp.restrictiveInstantiation
            );
        }
```
to align with the original intent - as is, since `constraint` is a lazily calculated field (and is calculated from the symbol), all this is doing right now is cloning the type parameter (constraint and all!) if it's constraint has been calculated already, or returning it as-is if it hasn't been calculated yet.

Fixing this, however, is not without warts. For example, take this type:
```ts
type IsDefinitelyDefined<T extends unknown> = [T] extends [{}] ? true : false;
```
when compared with an "unconstrained" `T` it's always _true_, but the `unknown` constraint makes it `false`. Maybe that's a good indicator that the constraint should be set as `unknownType` instead of `noConstraintType`?

Reply:
Agreed. The restrictive instantiation should be a type parameter with an `unknown` constraint.